File: construction_management/construction_management/report/dpr_register/dpr_register.js
frappe.query_reports["DPR Register"] = {
	onload: function () {
		apply_dpr_register_summary_cards();
	},
	refresh: function () {
		apply_dpr_register_summary_cards();
	},
	filters: [
		{
			fieldname: "project",
			label: __("Project"),
			fieldtype: "Link",
			options: "Project",
		},
		{
			fieldname: "customer",
			label: __("Customer"),
			fieldtype: "Link",
			options: "Customer",
		},
		{
			fieldname: "from_date",
			label: __("DPR Date From"),
			fieldtype: "Date",
		},
		{
			fieldname: "to_date",
			label: __("DPR Date To"),
			fieldtype: "Date",
		},
		{
			fieldname: "status",
			label: __("Status"),
			fieldtype: "Select",
			options: "\nDraft\nSubmitted\nPublished\nCancelled",
		},
		{
			fieldname: "publish_to_portal",
			label: __("Show on Client Portal"),
			fieldtype: "Check",
		},
	],
};

function apply_dpr_register_summary_cards() {
	ensure_dpr_register_summary_card_styles();

	[100, 500, 1000].forEach(function (delay) {
		setTimeout(tag_dpr_register_summary, delay);
	});
}

function tag_dpr_register_summary() {
	const $summary = $(".report-summary");
	if (!$summary.length) {
		return;
	}

	$summary.addClass("cm-dpr-register-summary");
}

function ensure_dpr_register_summary_card_styles() {
	if (document.getElementById("cm-dpr-register-summary-style")) {
		return;
	}

	const style = document.createElement("style");
	style.id = "cm-dpr-register-summary-style";
	style.textContent = `
		.report-summary.cm-dpr-register-summary {
			display: grid !important;
			grid-template-columns: repeat(3, minmax(0, 240px));
			gap: 32px;
			margin: 16px 0 24px;
			padding: 0;
			border: none;
			align-items: stretch;
			justify-content:space-around;
		}
		.report-summary.cm-dpr-register-summary .summary-item {
			display: flex;
			flex-direction: column;
			justify-content: space-between;
			width: 100%;
			max-width: none;
			min-height: 120px;
			height: auto;
			margin: 0;
			padding: 20px;
			border-radius: 8px;
			background: #ffffff;
			border: 1px solid #e5e7eb;
			border-left: 4px solid #d1d5db;
			box-shadow: 0 2px 8px rgba(0, 0, 0, 0.05);
			overflow: visible;
		}
		.report-summary.cm-dpr-register-summary .summary-label,
		.report-summary.cm-dpr-register-summary .summary-value {
			overflow: visible !important;
			text-overflow: clip !important;
			white-space: normal !important;
			overflow-wrap: anywhere;
		}
		@media (max-width: 991px) {
			.report-summary.cm-dpr-register-summary {
				grid-template-columns: repeat(2, minmax(0, 1fr));
			}
		}
		@media (max-width: 575px) {
			.report-summary.cm-dpr-register-summary {
				grid-template-columns: 1fr;
			}
		}
	`;
	document.head.appendChild(style);
}

File: construction_management/construction_management/report/dpr_register/dpr_register.py
import frappe
from frappe import _
from frappe.utils import cint


def execute(filters=None):
	filters = frappe._dict(filters or {})
	columns = get_columns()
	data = get_data(filters)
	return columns, data, None, None, get_report_summary(data)


def get_columns():
	return [
		{"label": _("DPR"), "fieldname": "name", "fieldtype": "Link", "options": "Daily Progress Report", "width": 170},
		{"label": _("DPR Date"), "fieldname": "dpr_date", "fieldtype": "Date", "width": 110},
		{"label": _("Project"), "fieldname": "project", "fieldtype": "Link", "options": "Project", "width": 170},
		{"label": _("Customer"), "fieldname": "customer", "fieldtype": "Link", "options": "Customer", "width": 170},
		{"label": _("Title"), "fieldname": "title", "fieldtype": "Data", "width": 220},
		{"label": _("Tasks Completed Count"), "fieldname": "tasks_completed_count", "fieldtype": "Int", "width": 160},
		{"label": _("Photos Count"), "fieldname": "photos_count", "fieldtype": "Int", "width": 110},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 100},
		{"label": _("Show on Client Portal"), "fieldname": "publish_to_portal", "fieldtype": "Check", "width": 160},
		{"label": _("Prepared By"), "fieldname": "prepared_by", "fieldtype": "Link", "options": "User", "width": 160},
		{"label": _("Modified"), "fieldname": "modified", "fieldtype": "Datetime", "width": 160},
	]


def get_data(filters):
	conditions = []
	values = {}

	if filters.get("project"):
		conditions.append("dpr.project = %(project)s")
		values["project"] = filters.project
	if filters.get("customer"):
		conditions.append("dpr.customer = %(customer)s")
		values["customer"] = filters.customer
	if filters.get("from_date"):
		conditions.append("dpr.dpr_date >= %(from_date)s")
		values["from_date"] = filters.from_date
	if filters.get("to_date"):
		conditions.append("dpr.dpr_date <= %(to_date)s")
		values["to_date"] = filters.to_date
	if filters.get("status"):
		conditions.append("dpr.status = %(status)s")
		values["status"] = filters.status
	if filters.get("publish_to_portal") not in (None, ""):
		conditions.append("dpr.publish_to_portal = %(publish_to_portal)s")
		values["publish_to_portal"] = cint(filters.publish_to_portal)

	where_clause = " AND ".join(conditions) if conditions else "1 = 1"

	return frappe.db.sql(
		f"""
		SELECT
			dpr.name,
			dpr.dpr_date,
			dpr.project,
			dpr.customer,
			dpr.title,
			COALESCE(tasks.total, 0) AS tasks_completed_count,
			COALESCE(photos.total, 0) AS photos_count,
			dpr.status,
			dpr.publish_to_portal,
			dpr.prepared_by,
			dpr.modified
		FROM `tabDaily Progress Report` dpr
		LEFT JOIN (
			SELECT parent, COUNT(*) AS total
			FROM `tabDPR Task Completed`
			WHERE parenttype = 'Daily Progress Report'
			  AND parentfield = 'tasks_completed'
			GROUP BY parent
		) tasks ON tasks.parent = dpr.name
		LEFT JOIN (
			SELECT parent, COUNT(*) AS total
			FROM `tabDPR Photo`
			WHERE parenttype = 'Daily Progress Report'
			  AND parentfield = 'photos'
			GROUP BY parent
		) photos ON photos.parent = dpr.name
		WHERE {where_clause}
		ORDER BY dpr.dpr_date DESC, dpr.modified DESC
		""",
		values,
		as_dict=True,
	)


def get_report_summary(data):
	return [
		{"value": len(data), "label": _("Total DPRs"), "datatype": "Int", "indicator": "Blue"},
		{
			"value": sum(
				1
				for row in data
				if row.status in ("Submitted", "Published") and cint(row.publish_to_portal)
			),
			"label": _("Published DPRs"),
			"datatype": "Int",
			"indicator": "Green",
		},
		{
			"value": sum(1 for row in data if row.status == "Draft"),
			"label": _("Draft DPRs"),
			"datatype": "Int",
			"indicator": "Gray",
		},
		{
			"value": sum(cint(row.tasks_completed_count) for row in data),
			"label": _("Total Task Entries"),
			"datatype": "Int",
			"indicator": "Blue",
		},
		{
			"value": sum(cint(row.photos_count) for row in data),
			"label": _("Total Photos"),
			"datatype": "Int",
			"indicator": "Orange",
		},
		{
			"value": len({row.project for row in data if row.project}),
			"label": _("Projects with DPRs"),
			"datatype": "Int",
			"indicator": "Green",
		},
	]
